python/ray/autoscaler/batching_node_provider.py

```diff
@@ -0,0 +1,210 @@
+        each node type to maintain.
+
+        Consistency requirement:
+        If a node id was present in ScaleRequest.workers_to_delete of a previously
+        submitted scale request, it should no longer be present as key in get_node_data.
+        (Node termination must be registered immediately when submit_scale_request
+        returns.)
+        """
+        raise NotImplementedError
+
+    def safe_to_scale() -> bool:
+        """Optional condition to determine if it's safe to submit a scale request.
+        Can be used to wait for convergence of state managed by an external cluster
+        manager.
+
+        If False is returned, the scale request will not be submitted during the current
+        autoscaler update iteration.
+        """
+        return True
+
+    def post_process(self) -> None:
+        """Submit a scale request if it is safe to do so."""
+        if self.scale_change_needed and self.safe_to_scale():
+            self.submit_scale_request(self.scale_request)
+        self.scale_change_needed = False
+
+    def non_terminated_nodes(self, tag_filters: Dict[str, str]) -> List[str]:
+        self.scale_change_needed = False
+        self.node_data_dict = self.get_node_data()
+
+        # Initialize ScaleRequest
+        self.scale_request = ScaleRequest(
+            desired_num_workers=self.cur_num_workers,  # Current scale
+            workers_to_delete=[],  # No workers to delete yet
+        )
+        return list(self.node_data_dict.keys())
+
+    @property
+    def cur_num_workers(self, node_data_dict: Dict[str, Any]):
+        num_workers_dict = defaultdict(int)
+        for node_data in node_data_dict.values():
+            node_type = node_data.node_tags[TAG_RAY_USER_NODE_TYPE]
+            num_workers_dict[node_type] += 1
+        return num_workers_dict
+
+    def node_tags(self, node_id: str) -> Dict[str, str]:
+        node_data = self.node_data_dict[node_id]
+        return {
+            TAG_RAY_NODE_KIND: node_data.kind,
+            TAG_RAY_NODE_STATUS: node_data.status,
+            TAG_RAY_USER_NODE_TYPE: node_data.type,
+        }
+
+    def internal_ip(self, node_id: str) -> str:
+        return self.node_data_dict[node_id].ip
+
+    def create_node(
+        self, node_config: Dict[str, Any], tags: Dict[str, str], count: int
+    ) -> Optional[Dict[str, Any]]:
+        node_type = tags[TAG_RAY_USER_NODE_TYPE]
+        self.scale_request.desired_num_workers[node_type] += count
+        self.scale_change_needed = True
+
+    def terminate_node(self, node_id: str) -> Optional[Dict[str, Any]]:
+        """Terminates the specified node.
+
+        Optionally return a mapping from deleted node ids to node
+        metadata.
+        """
+        node_type = self.node_tags(node_id)[TAG_RAY_USER_NODE_TYPE]
+        self.scale_request.desired_num_workers[node_type] -= 1
+        if self.scale_request.desired_num_workers[node_type] < 0:
```

Comment:
Shouldn't we add a check to keep `desired_num_workers` >= min_workers

Reply:
Yes. But
- The autoscaler already makes such a check. (We might end up at less than min due to an internal inconsistency, though.)
- Threading min and max worker data into the provider is probably doable, but requires additional code changes to the autoscaler (let me think about it)
- The KubeRay provider has direct access to the RayCr already, so it might be better to implement the logic in that subclass (on the other hand, there's some appeal to doing this in the abstract class)

Reply:
I'm going to implement that check in kuberaynodeprovider for the reasons above.

Reply:
Also, rather than raising here, perhaps should log a warning and ignore the change.